index.html

```diff
@@ -2971,15 +2971,35 @@ <h3>Add Proof Set/Chain</h3>
 the <var>unsecuredDocument</var>.
           </li>
           <li>
+Let <var>matchingProofs</var> be an empty set.
+          </li>
+          <li>
 If <var>options</var> contains a <code>previousProof</code> attribute and that
-attribute is a string, check whether
-an element of the <var>allProofs</var> has a matching <code>id</code> attribute.
-If not, an error MUST be raised and SHOULD convey an error type of
-<a href="#PROOF_GENERATION_ERROR">PROOF_GENERATION_ERROR</a>. If the
-<code>previousProof</code> attribute is an array, check that each element of that
-array is a string which matches the <code>id</code> attribute of an element
-of the <var>allProofs</var>. If not, an error MUST be raised and SHOULD convey
-an error type of <a href="#PROOF_GENERATION_ERROR">PROOF_GENERATION_ERROR</a>.
+attribute is a string, add the element from <var>allProofs</var> with a matching
+<code>id</code> attribute to matchingProofs. If it does not exist, an error MUST
+be raised and SHOULD convey an error type of
+<a href="#PROOF_GENERATION_ERROR">PROOF_GENERATION_ERROR</a>.
+If the <code>previousProof</code> attribute is an array, add an element from
+<var>allProofs</var> with an <code>id</code> attribute that matches each element
+of that array. If any matches are missing, an error MUST be raised and SHOULD
+convey an error type of
+<a href="#PROOF_GENERATION_ERROR">PROOF_GENERATION_ERROR</a>.
+          </li>
+          <li>
+For each element <var>mp</var> of <var>matchingProofs</var>, if there are any,
+recursively run the previous step on <var>mp.previousProof</var> if it exist.
+            <p class="note">
+This steps insures that all proof dependencies are added regardless of the depth
+of the proof chain.
```

Comment:
So I don't think we need to run this step recursively. This is because any proof dependency will have already signed over the contents of its own dependencies. For example, if we have a proof chain of A => B => C, then B will have signed over A such that when C is created, it only needs to sign over B, not A and B.

Reply:
Let me write up the security intent and considerations (attack model) for this first with some examples to be sure. If we don't it will be a nice simplification.

Reply:
Hi @dlongley, here is my example analysis and agrees with your notion that the recursive part is unneeded. Please take a look for holes or other attack scenarios.

## Proof Chains and Sets

Instead of a single *proof* (proof options with *proofValue*) included in the signed VC we can have an array of them. We call such a thing a *proof set*. Each one of which should verify with the appropriate keying material.

The *holder* (or intermediary) receiving a signed VC containing a *proof set* is able to remove any of the *proofs* within the set prior to passing it on to another entity.

We can introduce a "dependency relation" between *proofs* in a *proof set* by assigning the proof that another proof relies on, i.e., the *dependent proof*, an *id* and the *relying proof* will have a list (or single item) denoting *previousProofs*. Such *dependency chains* can have arbitrary depth. The **intent** of such a *proof chain* is to model an approval chain in a business process or a notary witnessing analog signatures. In such cases we CANNOT allow the *proofs* that are relied upon to be **substituted** with other proofs or removed.

Example: The CEO signed off on a PO (document), based on the reasonableness of the item being requested and the fact that the CFO and the VP of Engineering signed off on the PO. Suppose we have a proof set:

```javascript
proofs = [
  {id: 0, pubKey: pubVPEng,  previusProof: null, proofValue: signature(VP_Eng, unsignedDocument)},
  {id: 1, pubKey: pubCFO, previousProof: 0, proofValue: signature(privCFO, unsignedDocument)},
  {id: 2, pubKey: pubCEO, previousProof: 1, proofValue: signature(privCEO, unsignedDocument)}
]
```

Where *pubX* denotes a public key and *privX* denotes the corresponding private key, and `signature(privKey, doc)` denotes a signature algorithm.

A proof chain such as this provides **no** protection from the recipient swapping out *dependent* signatures as in the following example:

```javascript
[
  {id: 1, pubKey: pubHacker, previousProof: null, proofValue: signature(privHacker, unsignedDocument)},
  {id: 2, pubKey: pubCEO, previousProof: 2, proofValue: signature(privCEO, unsignedDocument)}
]
```

Hence in the chain case it is very important to "bind" the previous proofs in the chain such as in the following.

```javascript
[
  {id: 1, pubKey: pubVPEng, previousProof: null, proofValue: signature(privVPEng, unsignedDocument)},
  {id: 2, pubKey: pubCFO, previousProof: 1, proofValue: signature(privCFO, unsignedDocument + proofs[0])},
  {id: 3, pubKey: pubCEO, previousProof: 2, proofValue: signature(privCEO, unsignedDocument + proofs[1])}
]
```

Now when verifying we see that proof[0] depends on the proper signing by VPEng and integrity of the unsigned document. We see that proofs[1] depends on the proper signing by the CFO and the integrity of the unsigned document and proofs[0], i.e., binds `{id: 1, pubKey: pubVPEng, previousProof: null, proofValue: signature(VP_Eng, unsignedDocument)}` with the document so it cannot be removed or swapped out. For proofs[2] we see that it depends on the proper signing by the CEO and the integrity of the unsigned document and proofs[1] (which in turn depends on document and proofs[0]). Hence the above scheme with the additional binding guards the integrity of the proof chain.

Reply:
This analysis should probably be put somewhere it will be (easily) visible after this conversation is resolved.

Reply:
Where? With example near definition? Security considerations section?